[JSC] Add `DisposableStack` constructor from Explicit Resource Management Proposal

sosukesuzuki opened the pull request:
#### 6bf4c5606ebc30f7dc8e8537fbc18dd357bd9a2a
<pre>
[JSC] Add `DisposableStack` constructor from Explicit Resource Management Proposal
<a href="https://bugs.webkit.org/show_bug.cgi?id=292818">https://bugs.webkit.org/show_bug.cgi?id=292818</a>

Reviewed by Yusuke Suzuki.

This patch implements `DisposableStack` constructor[1] from Explicit
Resource Management Proposal[2].

I will implement the prototype methods in a subsequent PR.

[1]: <a href="https://tc39.es/proposal-explicit-resource-management/#sec-disposablestack">https://tc39.es/proposal-explicit-resource-management/#sec-disposablestack</a>
[2]: <a href="https://github.com/tc39/proposal-explicit-resource-management">https://github.com/tc39/proposal-explicit-resource-management</a>

* JSTests/test262/config.yaml:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapSubspaceTypes.h:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/DisposableStack.cpp: Added.
(JSC::DisposableStack::create):
(JSC::DisposableStack::finishCreation):
(JSC::DisposableStack::visitChildrenImpl):
* Source/JavaScriptCore/runtime/DisposableStack.h: Added.
* Source/JavaScriptCore/runtime/DisposableStackConstructor.cpp: Added.
(JSC::DisposableStackConstructor::createStructure):
(JSC::DisposableStackConstructor::create):
(JSC::DisposableStackConstructor::finishCreation):
(JSC::DisposableStackConstructor::visitChildrenImpl):
(JSC::DisposableStackConstructor::DisposableStackConstructor):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/DisposableStackConstructor.h: Added.
* Source/JavaScriptCore/runtime/DisposableStackConstructorInlines.h: Added.
* Source/JavaScriptCore/runtime/DisposableStackInlines.h: Added.
(JSC::DisposableStack::createStructure):
* Source/JavaScriptCore/runtime/DisposableStackPrototype.cpp: Added.
(JSC::DisposableStackPrototype::finishCreation):
* Source/JavaScriptCore/runtime/DisposableStackPrototype.h: Added.
* Source/JavaScriptCore/runtime/DisposableStackPrototypeInlines.h: Added.
(JSC::DisposableStackPrototype::createStructure):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::disposableStackStructure const):
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:

Canonical link: <a href="https://commits.webkit.org/294880@main">https://commits.webkit.org/294880@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/215bf3a15af3473b2ddc8e483f1e1524940d9fb8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/103407 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23089 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/13407 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/108581 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54051 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/23435 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/31640 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/78582 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/106413 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18163 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/93282 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/58917 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/18000 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/11327 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/53407 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/96082 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/87797 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/11388 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/110957 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/102018 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/30554 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/22547 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/87579 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/30915 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/89481 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87220 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22199 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32086 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/9801 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/24881 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/30482 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/35794 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/125651 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/30282 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/34832 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/33613 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/31843 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -28,6 +28,9 @@
 #include <optional>
 #include <wtf/Atomics.h>
+#if CPU(ARM_THUMB2)
+#include <wtf/MathExtras.h>
```
I'm not sure why, but without this change, the build fails on armv7 with an error saying that `WTF::roundUpToMultipleOf` cannot be found.
Since this file actually uses `WTF::roundUpToMultipleOf`, I think including this header file makes sense.

Constellation commented on Source/JavaScriptCore/assembler/AssemblerCommon.h:
```diff
@@ -28,6 +28,9 @@
 #include "OSCheck.h"
 #include <optional>
 #include <wtf/Atomics.h>
+#if CPU(ARM_THUMB2)
```
I don't think we should include the header with this guard. If it is necessary, then let's just include regardless of CPU device models.

Constellation reviewed (approved):
r=me with comment.

sosukesuzuki committed:
[JSC] Add `DisposableStack` constructor from Explicit Resource Management Proposal
https://bugs.webkit.org/show_bug.cgi?id=292818

Reviewed by Yusuke Suzuki.

This patch implements `DisposableStack` constructor[1] from Explicit
Resource Management Proposal[2].

I will implement the prototype methods in a subsequent PR.

[1]: https://tc39.es/proposal-explicit-resource-management/#sec-disposablestack
[2]: https://github.com/tc39/proposal-explicit-resource-management

* JSTests/test262/config.yaml:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/heap/Heap.h:
* Source/JavaScriptCore/heap/HeapSubspaceTypes.h:
* Source/JavaScriptCore/runtime/CommonIdentifiers.h:
* Source/JavaScriptCore/runtime/DisposableStack.cpp: Added.
(JSC::DisposableStack::create):
(JSC::DisposableStack::finishCreation):
(JSC::DisposableStack::visitChildrenImpl):
* Source/JavaScriptCore/runtime/DisposableStack.h: Added.
* Source/JavaScriptCore/runtime/DisposableStackConstructor.cpp: Added.
(JSC::DisposableStackConstructor::createStructure):
(JSC::DisposableStackConstructor::create):
(JSC::DisposableStackConstructor::finishCreation):
(JSC::DisposableStackConstructor::visitChildrenImpl):
(JSC::DisposableStackConstructor::DisposableStackConstructor):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/DisposableStackConstructor.h: Added.
* Source/JavaScriptCore/runtime/DisposableStackConstructorInlines.h: Added.
* Source/JavaScriptCore/runtime/DisposableStackInlines.h: Added.
(JSC::DisposableStack::createStructure):
* Source/JavaScriptCore/runtime/DisposableStackPrototype.cpp: Added.
(JSC::DisposableStackPrototype::finishCreation):
* Source/JavaScriptCore/runtime/DisposableStackPrototype.h: Added.
* Source/JavaScriptCore/runtime/DisposableStackPrototypeInlines.h: Added.
(JSC::DisposableStackPrototype::createStructure):
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::disposableStackStructure const):
* Source/JavaScriptCore/runtime/JSType.cpp:
(WTF::printInternal):
* Source/JavaScriptCore/runtime/JSType.h:

Canonical link: https://commits.webkit.org/294880@main
